subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultWorkPropertyMetadataStore.java

```diff
@@ -0,0 +1,191 @@
+import org.gradle.api.internal.tasks.properties.annotations.InputDirectoryPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.InputFilePropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.InputFilesPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.InputPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.LocalStatePropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.NestedBeanAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.NoOpPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.OutputDirectoriesPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.OutputDirectoryPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.OutputFilePropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.OutputFilesPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.OverridingPropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.annotations.PropertyAnnotationHandler;
+import org.gradle.api.internal.tasks.properties.bean.PropertyMetadata;
+import org.gradle.api.tasks.Console;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.SkipWhenEmpty;
+import org.gradle.api.tasks.options.OptionValues;
+import org.gradle.cache.internal.CrossBuildInMemoryCache;
+import org.gradle.cache.internal.CrossBuildInMemoryCacheFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DefaultWorkPropertyMetadataStore implements WorkPropertyMetadataStore {
+
+    private final static List<? extends PropertyAnnotationHandler> HANDLERS = Arrays.asList(
+        new InputFilePropertyAnnotationHandler(),
+        new InputDirectoryPropertyAnnotationHandler(),
+        new InputFilesPropertyAnnotationHandler(),
+        new OutputFilePropertyAnnotationHandler(),
+        new OutputFilesPropertyAnnotationHandler(),
+        new OutputDirectoryPropertyAnnotationHandler(),
+        new OutputDirectoriesPropertyAnnotationHandler(),
+        new InputPropertyAnnotationHandler(),
+        new DestroysPropertyAnnotationHandler(),
+        new LocalStatePropertyAnnotationHandler(),
+        new NestedBeanAnnotationHandler(),
+        new NoOpPropertyAnnotationHandler(Inject.class),
+        new NoOpPropertyAnnotationHandler(Console.class),
+        new NoOpPropertyAnnotationHandler(Internal.class),
+        new NoOpPropertyAnnotationHandler(OptionValues.class)
+    );
```

Comment:
I think this all should be in `:execution`, including the annotation handlers. Something like in `org.gradle.internal.execution.properties` or something.